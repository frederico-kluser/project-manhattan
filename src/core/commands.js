/* eslint-disable no-undef */
var key;

var activeCommands = false;
var shortCutCommands = {
  command: '',
  value: '',
  type: '',
};

const resetCommands = () => {
  activeCommands = false;
  shortCutCommands = {
    command: '',
    value: '',
    type: '',
  };
};

const executeCommands = () => {
  const {command, value} = shortCutCommands;

  // eslint-disable-next-line default-case
  switch (command) {
    case 'width':
    case 'height':
    case 'left':
    case 'top':
      Elements[elementId].style[command] = parseFloat(value);
      break;
  }

  updateGlobalStyle();
  resetCommands();
};

const getValue = () => {
  if (keyCodeCommands[key] !== undefined || keyCodeNumbers[key] !== undefined) {
    switch (keyCodeCommands[key]) {
      case 'backspace':
        shortCutCommands.value = backSpaceText(shortCutCommands.value);
        break;
      case 'enter':
        executeCommands();
        break;
      default:
        shortCutCommands.value += keyCodeNumbers[key];
        break;
    }
  } else {
    resetCommands();
  }
};

const setCommands = command => {
  shortCutCommands.command = command;
  return true;
};

const getCommandSwitch = {
  w: () => setCommands('width'),
  h: () => setCommands('height'),
  l: () => setCommands('left'),
  t: () => setCommands('top'),
  v: () => {
    recognition.continuous = false;
    recognition.start();
  },
};

const getCommands = e => {
  key = e.which || e.keyCode;
  console.log('key :', key);
  const letter = keyCodeLetters[key];

  if (activeCommands) {
    getValue();
  } else if (e.shiftKey) {
    activeCommands = !!dynamicFunction(getCommandSwitch[letter], resetCommands);
  }
};

document.body.addEventListener('keydown', getCommands, false);
