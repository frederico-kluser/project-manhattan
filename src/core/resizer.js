/* eslint-disable no-undef */
/* eslint-disable no-var */

var dragBegins = false;

var initialX;
var initialY;

var moveX;
var moveY;

var resizeX;
var resizeY;

var width = 200;
var height = 100;

const dragStart = e => {
  dragBegins = true;
  const pSize = e.target.children[0];
  pSize.innerHTML = `${width}px x ${height}px`;

  if (e.type === 'touchstart') {
    initialX = e.touches[0].layerX;
    initialY = e.touches[0].layerY;
  } else {
    initialX = e.layerX;
    initialY = e.layerY;
  }
};

const drag = e => {
  if (dragBegins) {
    if (e.type === 'touchstart') {
      moveX = e.touches[0].layerX;
      moveY = e.touches[0].layerY;
    } else {
      moveX = e.layerX;
      moveY = e.layerY;
    }

    resizeX = moveX - initialX;
    resizeY = moveY - initialY;

    const element = e.target;
    const pSize = element.children[0];
    pSize.innerHTML = `${width + resizeX}px x ${height + resizeY}px`;
    element.setAttribute('style', `width:${width + resizeX}px;height:${height + resizeY}px;`);
  }
};

const dragEnd = e => {
  dragBegins = false;
  width += resizeX;
  height += resizeY;

  const element = e.target;
  const pSize = element.children[0];
  console.log(element);
  console.log(`${width}px x ${height}px`);
  pSize.innerHTML = '';
};

const builderElement = () => {
  console.log('String to insert');
  const size = createElementor({tag: 'p', className: 'model-size'});
  const div = createElementor({tag: 'div', className: 'model', chield: [size]});
  div.addEventListener('mousedown', dragStart, false);
  div.addEventListener('mousemove', drag, false);
  div.addEventListener('mouseup', dragEnd, false);

  document.body.appendChild(div);
};

builderElement();
