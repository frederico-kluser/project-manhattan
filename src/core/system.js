/* eslint-disable no-undef */

var Elements = {};
const styleTemplate = {
  'background-color': 'lightblue',
  height: 200,
  left: 100,
  margin: 50,
  position: 'relative',
  top: 100,
  width: 300,
};
var elementId;

class ElementBuilder {
  constructor(parent, tag) {
    const {infoClass} = enums.helpers;
    const {sizeMode} = enums.mod;

    this.id = unicGlobalVarNameGenerator();
    elementId = this.id;
    this.info = createElement({tag: 'span', className: infoClass});
    this.mode = sizeMode;
    this.style = styleTemplate;
    this.tag = tag;

    this.dragBegins = false;

    this.initialPositionX = 0;
    this.initialPositionY = 0;

    this.positionX = 0;
    this.positionY = 0;

    this.calcValueX = 0;
    this.calcValueY = 0;

    this.element = createElement({
      tag: this.tag,
      chield: [this.info],
      id: this.id,
    });

    this.element.addEventListener('mousedown', dragStart, false);
    this.element.addEventListener('mousemove', dragMove, false);
    this.element.addEventListener('mouseup', dragEnd, false);
    this.element.addEventListener('contextmenu', builderMenu);

    parent.appendChild(this.element);
    Elements[this.id] = this;
    if (globalStyle === undefined) {
      injectCSS();
    }
  }

  get getElement() {
    return this.element;
  }

  setInfo(type) {
    const {sizeMode, moveMode} = enums.mod;
    const {width, height, left, top} = this.style;
    let info = '';

    switch (this.mode) {
      case sizeMode:
      default:
        info = `${height + this.calcValueY}px X ${width + this.calcValueX}px`;
        break;
      case moveMode:
        info = `${top + this.calcValueY}px X ${left + this.calcValueX}px`;
        break;
    }

    if (type === 'move' && !this.dragBegins) {
      info = '';
    }
    this.info.innerHTML = info;
  }

  setStyleTagAttribute() {
    const {sizeMode, moveMode} = enums.mod;
    const {width, height, left, top} = this.style;
    let style = '';

    switch (this.mode) {
      case sizeMode:
      default:
        style = `width:${width + this.calcValueX}px;height:${height + this.calcValueY}px;`;
        break;
      case moveMode:
        style = `left:${left + this.calcValueX}px;top:${top + this.calcValueY}px;`;
        break;
    }

    this.element.setAttribute('style', style);
  }

  updateStyleElement() {
    const {sizeMode, moveMode} = enums.mod;
    switch (this.mode) {
      case sizeMode:
      default:
        this.style.width += this.calcValueX;
        this.style.height += this.calcValueY;
        break;
      case moveMode:
        this.style.left += this.calcValueX;
        this.style.top += this.calcValueY;
        break;
    }
  }

  drag(e, type) {
    const x = e.clientX;
    const y = e.clientY;

    switch (type) {
      case 'start':
        elementId = this.id;
        this.dragBegins = true;

        this.initialPositionX = x;
        this.initialPositionY = y;

        this.calcValueX = 0;
        this.calcValueY = 0;
        break;
      case 'move':
        if (this.dragBegins) {
          this.PositionX = x;
          this.PositionY = y;

          this.calcValueX = this.PositionX - this.initialPositionX;
          this.calcValueY = this.PositionY - this.initialPositionY;

          this.setStyleTagAttribute();
        }
        break;
      case 'end':
      default:
        this.dragBegins = false;

        this.updateStyleElement();

        this.calcValueX = 0;
        this.calcValueY = 0;

        updateGlobalStyle();
        break;
    }

    this.setInfo(type);
  }
}

// eslint-disable-next-line no-new
new ElementBuilder(document.body, 'div');
window.elementId = elementId;
