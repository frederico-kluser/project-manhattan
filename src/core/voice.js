/* eslint-disable no-undef */
/* eslint-disable no-use-before-define */

var SpeechRecognition = SpeechRecognition || webkitSpeechRecognition;
var SpeechGrammarList = SpeechGrammarList || webkitSpeechGrammarList;
var SpeechRecognitionEvent = SpeechRecognitionEvent || webkitSpeechRecognitionEvent;

var recognition = new SpeechRecognition();
var speechRecognitionList = new SpeechGrammarList();
var grammar = (type, arr) => `#JSGF V1.0; grammar ${type}; public <${type}> = ${arr.join(' | ')} ;`;

const commandsObject = {
  directions: ['esquerda', 'direita', 'acima', 'abaixo'],
  proportion: ['largura', 'altura'],
  sizes: ['porcentagem', 'pixels', 'largura', 'altura'],
};
var voiceCommands = [];

Object.values(commandsObject).forEach(words => {
  voiceCommands = voiceCommands.concat(words);
});

speechRecognitionList.addFromString(grammar('commands', voiceCommands), 1);

recognition.grammars = speechRecognitionList;
recognition.continuous = false;
recognition.lang = 'pt-BR';
recognition.interimResults = false;
recognition.maxAlternatives = 5;

// recognition.start();

recognition.onresult = function (event) {
  const lastResult = event.results.length - 1;
  const results = event.results[lastResult];
  const possiblesOutput = [];
  console.log(`Result received: ${results[0].transcript}`);
  console.log(`Confidence: ${results[0].confidence}`);
  Object.values(results).forEach(({transcript}) => {
    possiblesOutput.push(transcript);
  });
  console.log(possiblesOutput.join(', '));
};

export default recognition;
