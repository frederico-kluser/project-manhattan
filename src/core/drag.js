import {elementIdSetter, Elements} from './elements.js';

export const dragStartEventSetter = e => {
  if (e.which === 1) {
    const {id} = e.target;
    elementIdSetter(id);
    Elements[id].dragEvents(e, 'start');
  }
};

export const dragMoveEventSetter = e => {
  const {id} = e.target;
  Elements[id].dragEvents(e, 'move');
};

export const dragEndEventSetter = e => {
  const {id} = e.target;
  Elements[id].dragEvents(e, 'end');
};
