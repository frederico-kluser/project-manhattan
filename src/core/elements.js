import {enums} from '../helpers/enums.js';
import {
  builderMenu,
  createElement,
  getRandomColor,
  unicGlobalVarNameGenerator,
} from '../helpers/general.js';
import regex from '../helpers/regex.js';
import {propertySizeInPixels} from '../styles/cssProperties.js';
import {
  makeOnlyElementtouchable,
  styleTagUpdater,
  updateHelperBubble,
  fixPropertySize,
} from '../styles/styleGeneral.js';
import {dragEndEventSetter, dragMoveEventSetter, dragStartEventSetter} from './drag.js';

export const Elements = {};

const stylePropertyDefault = () => ({
  border: '1px solid black',
  height: 100,
  left: 100,
  position: 'relative',
  top: 100,
  width: 100,
});

let elementId;
export const elementIdGetter = () => elementId;
export const elementIdSetter = value => {
  elementId = value;
};

export const ElementConfig = (id, element) => {
  const {sizeMode} = enums.mod;
  const {backgroundColor} = enums.style;

  elementIdSetter(id);
  let _mode = sizeMode;
  const _style = stylePropertyDefault();
  _style[backgroundColor] = getRandomColor(500);

  let _onDraggingElement = false;

  let _initialPositionX = 0;
  let _initialPositionY = 0;

  let _positionX = 0;
  let _positionY = 0;

  let _calcValueX = 0;
  let _calcValueY = 0;

  const _element = element;

  const _styleDynamic = {};
  Object.keys(propertySizeInPixels).forEach(property => {
    _styleDynamic[`${property}Absolute`] = _element[propertySizeInPixels[property]];
    _styleDynamic[`${property}Relative`] = _style[property] || 0;
  });

  const _sizeConverter = (type, plus, resumeOutput = false) => {
    // eslint-disable-next-line no-restricted-globals
    const sizeType = isNaN(_styleDynamic[`${type}Relative`])
      ? regex.getSizeType(_styleDynamic[`${type}Relative`])
      : false;
    if (sizeType !== false) {
      const valueType = parseFloat(_styleDynamic[`${type}Relative`]);

      const absoluteSizeBefore = _styleDynamic[`${type}Absolute`];
      const absoluteSizeAfter = absoluteSizeBefore + plus;

      const relativeSizeAfter = (absoluteSizeAfter * valueType) / absoluteSizeBefore;

      return (resumeOutput ? relativeSizeAfter.toFixed(2) : relativeSizeAfter) + sizeType;
    }
    return _style[type] + (plus || '');
  };

  const _helperBubbleTextRender = () => {
    const {moveMode} = enums.mod;
    let output;
    const resumeOutput = true;

    switch (_mode) {
      case sizeMode:
      default:
        output =
          fixPropertySize(_sizeConverter('height', _calcValueY, resumeOutput), 'height', true) +
          fixPropertySize(_sizeConverter('width', _calcValueX, resumeOutput), 'width', true);
        break;
      case moveMode:
        output =
          fixPropertySize(_sizeConverter('top', _calcValueY, resumeOutput), 'top', true) +
          fixPropertySize(_sizeConverter('left', _calcValueX, resumeOutput), 'left', true);
        break;
    }

    return output;
  };

  const _dynamicStylePropertiesTextSetter = () => {
    const {tune, search} = enums.icons;
    let info = _helperBubbleTextRender();
    let icon = tune;

    if (!_onDraggingElement) {
      info = '';
      icon = search;
    }

    updateHelperBubble(info, icon);
  };

  const _fixSizeTypes = property => {
    let output;

    if (regex.test(_style[property], regex.onlyNumbers)) {
      output = parseFloat(_style[property]);
    } else {
      output = _style[property];
    }

    _style[property] = output;
  };

  const _updateDynamicProperties = () => {
    const styleAttributeHTML = _helperBubbleTextRender();

    _fixSizeTypes('height');
    _fixSizeTypes('width');
    _fixSizeTypes('top');
    _fixSizeTypes('left');

    _element.setAttribute('style', styleAttributeHTML);
  };

  const _updateElementStyles = () => {
    const {moveMode} = enums.mod;

    switch (_mode) {
      case sizeMode:
      default:
        _style.width = _sizeConverter('width', _calcValueX);
        _style.height = _sizeConverter('height', _calcValueY);
        break;
      case moveMode:
        _style.left = _sizeConverter('left', _calcValueX);
        _style.top = _sizeConverter('top', _calcValueY);
        break;
    }
  };

  return {
    elementGetter: () => _element,
    dragEvents: (e, type) => {
      const x = e.clientX;
      const y = e.clientY;

      switch (type) {
        case 'start':
          _onDraggingElement = true;

          _initialPositionX = x;
          _initialPositionY = y;

          _calcValueX = 0;
          _calcValueY = 0;

          Object.keys(propertySizeInPixels).forEach(property => {
            _styleDynamic[`${property}Absolute`] = _element[propertySizeInPixels[property]];
            _styleDynamic[`${property}Relative`] = _style[property] || 0;
          });

          styleTagUpdater();
          makeOnlyElementtouchable(_element);
          break;
        case 'move':
          if (_onDraggingElement) {
            _positionX = x;
            _positionY = y;

            _calcValueX = _positionX - _initialPositionX;
            _calcValueY = _positionY - _initialPositionY;

            _updateDynamicProperties();
          }
          break;
        case 'end':
          _onDraggingElement = false;

          _updateElementStyles();

          _calcValueX = 0;
          _calcValueY = 0;

          styleTagUpdater();
          break;
      }

      _dynamicStylePropertiesTextSetter();
    },
    modeSetter: mode => {
      _mode = mode;
    },
    styleGetter: () => _style,
    styleSetter: (command, value) => {
      _style[command] = value;
    },
  };
};

export const ElementBuilder = (parent, tag, className = '', text = '') => {
  const id = unicGlobalVarNameGenerator();
  const element = createElement({className, id, tag, text});
  element.addEventListener('mousedown', dragStartEventSetter);
  element.addEventListener('mousemove', dragMoveEventSetter);
  element.addEventListener('mouseup', dragEndEventSetter);
  element.addEventListener('contextmenu', builderMenu);
  parent.appendChild(element);

  Elements[id] = ElementConfig(id, element);
};
