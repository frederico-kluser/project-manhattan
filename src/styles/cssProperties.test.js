import {ElementBuilder} from '../core/elements';
import {
  propertyCommands,
  propertyHasSizeType,
  propertyShiftCommands,
  propertySizeInPixels,
} from './cssProperties';

describe('object propertyShiftCommands', () => {
  it('default value', () => {
    expect.assertions(3);

    expect(propertyShiftCommands.h).not.toBeUndefined();
    expect(propertyShiftCommands.h()).not.toBeUndefined();
    expect(propertyShiftCommands).toMatchSnapshot();
  });
});

describe('object propertyHasSizeType', () => {
  it('default value', () => {
    expect.assertions(1);

    expect(propertyHasSizeType).toStrictEqual({
      padding: true,
      margin: true,
      top: true,
      left: true,
      width: true,
      height: true,
    });
  });
});

describe('object propertyCommands', () => {
  ElementBuilder(document.body, 'div');

  it('default value', () => {
    expect.assertions(6);

    expect(propertyCommands.height({command: 'height', value: ''})).toBeTruthy();
    expect(propertyCommands.left({command: 'left', value: ''})).toBeTruthy();
    expect(propertyCommands.margin({command: 'margin', value: ''})).toBeTruthy();
    expect(propertyCommands.padding({command: 'padding', value: ''})).toBeTruthy();
    expect(propertyCommands.top({command: 'top', value: ''})).toBeTruthy();
    expect(propertyCommands.width({command: 'width', value: ''})).toBeTruthy();
  });
});

describe('object propertySizeInPixels', () => {
  it('default value', () => {
    expect.assertions(1);

    expect(propertySizeInPixels).toStrictEqual({
      top: 'offsetTop',
      left: 'offsetLeft',
      width: 'clientWidth',
      height: 'clientHeight',
    });
  });
});
