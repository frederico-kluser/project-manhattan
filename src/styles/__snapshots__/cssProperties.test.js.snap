// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object propertyShiftCommands default value 1`] = `
Object {
  "h": [Function],
  "l": [Function],
  "m": [Function],
  "p": [Function],
  "t": [Function],
  "w": [Function],
}
`;
